File: src/hello_world_cpp/src/publisher.cpp
#include <chrono>

#include "rclcpp/rclcpp.hpp"
#include "std_msgs/msg/string.hpp"

using namespace std::chrono_literals;

// 继承于rclcpp::Node的自定义节点HelloWorldPublisher
class HelloWorldPublisher : public rclcpp::Node
{
public:
    HelloWorldPublisher()
        : Node("publisher") // 构造时指定节点的名字
        , m_count(0)
    {
        m_publisher = this->create_publisher<std_msgs::msg::String>("/hello_world", 10);
        m_timer = this->create_wall_timer(500ms, std::bind(&HelloWorldPublisher::timercallback, this));
    }

private:
    void timercallback()
    {
        auto message = std_msgs::msg::String();
        message.data = "Hello, world! " + std::to_string(this->m_count++);
        RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", message.data.c_str());
        this->m_publisher->publish(message);
    }

private:
    rclcpp::TimerBase::SharedPtr m_timer;
    rclcpp::Publisher<std_msgs::msg::String>::SharedPtr m_publisher;
    size_t m_count;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<HelloWorldPublisher>();
    rclcpp::spin(node); // 使节点循环运行
    rclcpp::shutdown();
    return 0;
}

File: src/hello_world_py/hello_world_py/publisher.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import String


class HelloWorldPublisher(Node):
    def __init__(self):
        super().__init__('publisher')
        self.count_ = 0
        self.publisher_ = self.create_publisher(String, '/hello_world', 10)
        timer_period = 1.0  # seconds
        self.timer_ = self.create_timer(timer_period, self.timer_callback)

    def timer_callback(self):
        msg = String()
        msg.data = f"Hello, world! {self.count_}"
        self.get_logger().info(f"Publishing: '{msg.data}'")
        self.publisher_.publish(msg)
        self.count_ += 1


def main(args=None):
    rclpy.init(args=args)
    hello_world_publisher_node = HelloWorldPublisher()
    rclpy.spin(hello_world_publisher_node) # 启动节点的事件循环
    hello_world_publisher_node.destroy_node() # 清理并关闭节点
    rclpy.shutdown() # 关闭ROS2


if __name__ == '__main__':
    main()